Java를 활용한 Todo REST API 구현

MJbae opened the pull request:
안녕하세요, 트레이너님 

1주차 과제, 열심히 해보겠습니다!!

감사합니다.

MJbae committed:
Define Webserver

MJbae committed:
Define initial App

MJbae committed:
Define initial HttpHandler

MJbae committed:
Remove unused Test

MJbae committed:
Define port number and back log as constant

MJbae committed:
Define HttpMethod Enum

MJbae committed:
Define method of getting HttpMethod type var

MJbae committed:
Adjust getHttpMethod in DemoHttpHandler::handle

hannut91 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,25 @@
+        }
     }
-}
+}
```
<img width="204" alt="image" src="https://user-images.githubusercontent.com/14071105/166392678-ce4bd779-076f-4d45-96ff-588b326cc957.png">


코드 마지막 줄에는 빈 줄이 하나가 추가되어야 합니다. 그렇지 않으면 위와 같이 GitHub에서 표시가 됩니다.

이 것이 문제가 되는 이유는 POSIX 표준에서 Line을 `개행 문자로 끝나는 0개 이상의 개행 문자가 아닌 문자들의 시퀀스`으로 정의하고 있기 때문입니다.

```jsx
// Hello
```

만약 위와 같이 빈 줄이 없는 코드에서 아래에 코드를 추가하게 되면 Git은 다음과 같이 표시합니다.

```jsx
- // Hello
+ // Hello
+ // World
```

저는 단순히 아래 한 줄만 추가했는데 왜 세줄이 변경되었다고 나오는 걸까요? 그 이유는 줄바꿈 문자 `\n`가 들어갔기 때문입니다.

```jsx
- // Hello
+ // Hello\n
+ // World
```

만약 미리 빈 줄을 하나 추가한 상태에서 아래에 새로운 코드를 작성할 경우에는 다음과 같이 보여집니다.

```jsx
+ // World
```

이미 `\n`이 추가되어 있기 때문이죠. 따라서 코드 마지막 줄에 빈 줄 하나를 항상 추가해주는게 좋습니다.

혹은 자동으로 추가할 수 있도록 설정하는 방법도 있습니다. 종립님의 블로그 글을 참고해 주세요

See also
* [파일-마지막에-자동으로-1개의-개행-문자-추가하기](https://johngrib.github.io/wiki/intellij/#파일-마지막에-자동으로-1개의-개행-문자-추가하기)

MJbae replied:
아무래도 전 게으르니 자동으로 빈 줄 추가하는 설정을 해야겠네요! 참고 자료도 감사합니다~

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
+        HttpMethod method = getHttpMethod(exchange);
```
이 함수를 사용하는 입장에서는 이 함수가 `null`을 반환할 수 있는지 예상할 수 없고, 그리고 `null`처리를 하지 않아도 자바 컴파일러는 아무런 불만도 표현하지 않기 때문에 위험성이 있는 것 같아요. 만약 반환 타입을 Optional 그대로 반환한다면, 함수의 Signature에서 이 값이 있을 수도 있고 없을 수도 있다는 의도를 드러내고 있고, 없는 경우를 처리하지 않으면 컴파일도 되지 않아서 에러 처리를 강제할 수 있어요

하지만 Optional 반환은 다른 부분이 어떻게 만들어졌냐가 관건이라고 보는데, 함수형으로 처리 중이 아니라면 예외가 가장 일반적이죠. Optional을 리턴하면 사용하는 쪽에서 비슷한 코드를 작성해야 하기 때문에 사용하기 힘든 인터페이스가 될 가능성이 커요.

MJbae replied:
이렇게 좋은 리뷰라니! 감사합니다.
`null`처리가 부재한 `getHttpMethod`에 대해 `Optional 타입 사용`과 `예외 던지기`라는 두 가지 대안을 제시해주신 것으로 이해했습니다.
말씀해주신대로 `Optional 타입 사용`은 비용이 크다고 판단하여 `예외 던지기`를 사용해서 수정해보겠습니다.

MJbae committed:
fix: null 처리를 위해 DemoHttpHandler::getHttpMethod 수정

why: 수정 전에는 null을 반환해도 컴파일러에서 걸러내지 못해서 소프트웨어의 견고함이 떨어졌음
what: 메소드에서 결과 값을 반환하기 직전에 null인 경우 예외를 던지는 로직 추가

MJbae committed:
refactor: 코드의 간결함을 위해 Request Path를 추출하는 로직을 별도 메소드로 추출

MJbae committed:
fix: null 처리 위해 ttpExchange에 예외 던지기 추가

why: 컴파일링 단계에서 null로 반환되는 로직을 쉽게 찾아내기 위함
what: uri, path 반환값이 null인 겨우 IllegalArgumentException 예외 던지기

MJbae committed:
refactor: handle 메소드 내 getHttpRequestPath 메소드 적용

MJbae committed:
fix: 예외 처리 위해 내부 메소드에서 던지는 예외를 상위 메소드로 전달

MJbae commented:
@hannut91 
윤석님의 리뷰를 받고 생각해보니, commit 메시지는 협력하는 동료 개발자에게 
작업 결과물을 쉽게 이해하기 위해 존재한다는 생각이 들었습니다.
그래서 commit 메시지 제목에 why와 what을 축약해서 담아보았습니다. 
구체적인 설명이 필요한 커밋의 경우에는 본문에 내용을 추가했습니다!

예외메시지는 빠른 디버깅을 위해 존재한다고 생각해서, 에러메시지에 예외발생 원인을 담았습니다.

MJbae committed:
clean: POSIX 표준을 지키기 위해 파일 끝 개행문자 추가

MJbae committed:
clean: 코드의 명시성을 높이기 위해 불필요한 import문 제거

MJbae committed:
feat: POST 요청을 처리하기 위해 getHttpRequestBody 메소드 추가

MJbae committed:
feat: POST 요청에 대해 분기처리하는 로직 추가

MJbae committed:
feat: Request Path 내 taskId를 추출하기 위한 메소드 추가

MJbae committed:
feat: PATCH 요청을 분기처리하기 위한 로직 추가

MJbae committed:
feat: DELETE 요청을 분기처리하기 위한 로직 추가

MJbae committed:
feat: 읽기 쉬운 코드를 만들기 위해 http status code를 상수로 정의

MJbae committed:
feat: HTTP method 마다 Response Header를 전달하기 위해 분기처리된 구간마다 sendResponseHeaders 호출

MJbae committed:
refactor: 코드의 간결성을 위해 Http Response를 보내는 로직을 메소드로 추출

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,61 @@
+    private String getHttpRequestPath(HttpExchange exchange) {
+        // null 처리 분기 추가
+        URI uri = exchange.getRequestURI();
+        if (uri == null) throw new IllegalArgumentException("failed to return URI");
```
한 줄이더라도 중괄호를 사용하시는 걸 권해드립니다. 물론 팀의 컨벤션의 따르는게 정답이긴 하지만요

MJbae replied:
그렇군요! 말씀듣고 찾아보니 코드의 일관성면에서 한 줄이라도 중괄호를 쓰는 것 선호하는 경향이 많은 것 같습니다.

MJbae replied:
> hannut91



> 



> 

코드 리뷰는 양방향의 상호작용인데, 제가 너무 일방적으로만 생각한 경향이 있네요 ㅎㅎ 
동등한 입장에서 코드에 대한 좋은 의견을 주고 받아보시죠~

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,61 @@
+        outputStream.close();
+    }
+
+    private String getHttpRequestPath(HttpExchange exchange) {
```
앞에 `get`을 붙이지 않아도 괜찮을 것 같아요. 우리가 원하는 것을 명시하는 것만으로도 충분히 의도가 전달 될 것 같습니다.

MJbae replied:
@hannut91 윤석님 말씀처럼 get을 붙이지 않아도 의도 전달은 가능하다는데 동의합니다~
그런데 Java naming convention을 살펴보니 메소드의 이름은 동사로 짓는 것을 권장하는 것 같습니다~
참고: https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

hannut91 replied:
아샬님이 남기신 것을 보셨을 수도 있지만, 객체지향 세상에서는 그저 원하는 것을 명시할 뿐, 어떻게 가져오는지 명시하지 않는게 좋다고 생각합니다

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,61 @@
+            content = "content with GET";
+        }
+
+        exchange.sendResponseHeaders(200, content.getBytes().length);
```
200이라는 숫자도 특별한 의미를 가지는 것이니 상수를 사용하는게 좋아보여요

See also
  - https://refactoring.com/catalog/replaceMagicLiteral.html

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,61 @@
+        return path;
+    }
+
+    private HttpMethod getHttpMethod(HttpExchange exchange) {
+        HttpMethod result = null;
+        String methodInString = exchange.getRequestMethod();
+
+        for (HttpMethod method : HttpMethod.values()) {
+            if (method.toString().equalsIgnoreCase(methodInString)) {
+                result = method;
+                break;
+            }
+        }
+
+        if (result == null) throw new IllegalArgumentException("failed to return HttpMethod enum value");
+
+        return result;
+    }
```
```java
private HttpMethod getHttpMethod(HttpExchange exchange) {
    String methodInString = exchange.getRequestMethod();
    return Arrays.stream(HttpMethod.values())
            .filter((method) ->
                 method.toString().equalsIgnoreCase(methodInString)
            )
            .findFirst()
            .orElseThrow(() ->
                    new IllegalArgumentException(
                            "failed to return HttpMethod enum value"
                    )
            );
}
```

스트림을 사용하는 것도 고려해 보세요

MJbae replied:
스트림 사용을 고려해봤지만, 스트림 체인으로 로직 연결에 실패했는데 이런 방법이 있었군요!
스트림에 대해 좀더 공부해봐야겠네요!! 감사합니다~

MJbae committed:
feat: 요구사항의 Task 데이터를 다루기 위해 Task 모델 정의

what: Task 인스턴스 객체 생성 시 Task 클래스 멤버 변수인 maxId를 기준으로 1을 증가신킨 값을 Id로 부여

MJbae committed:
feat: Task 객체를 JSON 포맷으로 변환하기 위한 메소드 정의

MJbae committed:
refactor: 구조의 유연함을 위해 tasks를 외부에서 주입 받도록 수정

MJbae committed:
refactor: 코드의 가독성을 높이기 위해 스트림 활용하여 개선

MJbae committed:
refactor: 코드의 일관성을 지키기 위해 중괄호 포함

MJbae committed:
clean: 불필요한 주석 제거

MJbae commented:
안녕하세요, 트레이너님
이번 작업에서는 "GET tasks/" 및 "POST tasks/"요청에 대한 테스트를 통과하는 것을 목표로 작업했습니다.
- 주요 작업은 Task 모델을 정의하는 부분이었습니다. 특히, Task 모델의 인스턴스 생성 시 ID가 자동생성되는 부분에 신경썼습니다.
- 구조의 유연함을 위해 List<Task> tasks는 최상위 계층인 App에서 주입 받도록 작업했습니다.
- 이외에도 리뷰로 남겨주신 부분들 대부분 반영하였습니다 :)

~~이번에도 좋은 리뷰 기대하겠습니다~~

감사합니다 :)

MJbae committed:
refactor: 코드의 간결성을 높이기 위해 후위연산자 활용

MJbae committed:
fix: jackson의 ObjectMapper 활용에 따른 역직렬화를 위해 기본생성자 추가

MJbae committed:
feat: Task 객체 인스턴스를 직렬화하기 위해 taskToJson 메소드 추가

MJbae committed:
refactor: 불필요한 객체 중복 생성 피하기 위해 외부에서 ObjectMapper를 매개변수로 전달

MJbae committed:
feat: json 형태의 데이터를 Task 객체로 역직렬화하기 위한 메소드 생성

hannut91 commented:
`이번에도 좋은 리뷰 기대하겠습니다~`라는 말씀은 조금 아쉽습니다. 물론 코드숨에서는 돈을 지불받고, 서비스를 제공하는 입장이기 때문에 그럴 수 있지만, 코드 리뷰라는 것이 특정 누군가가 누군가에게 제공하는 것은 아닙니다. 만약 이렇게 생각하신다면, 나는 실력이 안되서 리뷰를 못해, 나보다 잘하는 사람한테는 리뷰를 할 수가없어 같은 일들이 생깁니다.

코드 리뷰는 서로 상호간의 의견을 주고 받는 행위입니다. 트레이너가 같은 팀원이라고 생각하고, 서로 많은 의견을 주고받도록 해보세요

hannut91 commented on app/src/main/java/com/codesoom/assignment/Task.java:
```diff
@@ -0,0 +1,38 @@
+        return title;
+    }
+
+    @Override
+    public String toString() {
+        return "Task{" +
+                "id=" + id +
+                ", title='" + title + '\'' +
+                '}';
+    }
```
String.format을 사용해서 표현하는 건 어떨까요? 다음과 같이 표현할 수 있을거에요

```java
return String.format("{ id = %s, title = %s }", id, title);
```

더 좋은건 Template literal이지만 자바에서는 지원하질 않네요 ㅠ 만약 이런게 지원하는 코틀린을 사용했다면 다음과 같이 사용할 수 있을거에요

MJbae replied:
한 눈에 더 잘 들어오네요~ IDE의 추천으로 자동 완성된 것 보다 더 좋은 것 같습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,131 @@
+        String path = getHttpRequestPath(exchange);
+        String body = getHttpRequestBody(exchange);
+
+        if (method == GET && path.equals("/tasks")) {
+            content = tasksToJson(mapper);
+            exchange.sendResponseHeaders(HTTP_OK_CODE, content.getBytes().length);
+        }
```
조건문이 길어지다보니, 앞의 조건에 부합한다고 하더라도, 함수의 진행이 끝나지 않고 계속 되다보니, 사실은 필요없는 코드들도 계속 읽어야 할 수 밖에 없었어요.

만약 ~한 조건일 때 필요한 일만하고 빠르게 함수를 종료한다면, 내가 필요한 코드만 빠르게 읽을 수 있을 것 같아요.

MJbae replied:
조건문 이하의 로직에 대해 메소드로 분리하면 더 좋겠군요! 이렇게 되면 코드의 가독성이 더 높아지겠네요. 감사합니다~

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,131 @@
+        }
+
+        if (method == DELETE && path.startsWith("/tasks")) {
+            int taskId = extractTaskIdFromPath(path);
```
`extractTaskIdFromPath(path)` path가 두 번 반복되는 것이 어색해 보이네요. 메서드 이름에 `Path`가 없다면 어떻게 될까요?

MJbae replied:
메소드 행위의 의미는 그대로 전달되지만 사이즈는 줄어드니 더 좋네요~

MJbae replied:
> hannut91



> 



> 

코드 리뷰는 양방향의 상호작용인데, 제가 너무 일방적으로만 생각한 경향이 있네요 ㅎㅎ 
동등한 입장에서 코드에 대한 좋은 의견을 주고 받아보시죠~

MJbae committed:
refactor: Task 조회 작업의 시간복잡도를 줄이기 위해 자료구조 변경(List -> Map)

이유
- List 내부요소 조회 시 시간 복잡도가 O(n)이지만 Map에서 키를 기준으로 조회하면 O(1)로 시간복잡도가 줄어듬

작업내용
- List<Task>를 Map(Long, Task>로 변경
- values() 활용하여 GET 요청에 따른 response body 구성
- generateId 메소들르 Task 객체 내부에서 DemoHttpHandler로 이동

MJbae committed:
feat: PATCH 요청에 따라 기존 Task를 찾아서 새로운 Task로 변경하는 로직 추가

작업내용
- path param으로 전달 받은 taskId로 원하는 Task 찾는 로지 구현
- request body로 전달 받은 내용을 바탕으로 새로운 Task 객체 생성
- 기존 Task의 id를 새로운 Task 객체에 주입
- Map<Long, Task> tasks에서 기존 Task를 새로 생성한 Task로 대체

MJbae committed:
feat: DELETE 요청을 처리하기 위해 taskId에 해당하는 Task를 삭제하는 로직 구현

작업내용
- request param으로 전달 받은 taskId에 해당하는 Task를 Map<Long, Task> tasks에서 제거

MJbae committed:
fix: taskId에 해당하는 Task가 없는 경우 처리

작업 내용
- Map<Long, Task>에서 Key에 해당하는 Task가 없는 경우, Response Header에 404 코드 전달

MJbae committed:
feat: PUT 요청을 처리하는 로직 구현

- PATCH 요청에 따라 Task를 수정하는 로직과 동일

MJbae committed:
refactor: OutputStream 생성 후 예외 발생에 따른 close 실패 방지를 위해 try-catch-finally 적용

MJbae committed:
refactor: 중복되는 sendResponseHeaders 메소드를 sendResponse 내로 이동

MJbae committed:
feat: taskId로 Task 조회 요청을 처리하는 로직 구현

작업내용
- GET 메소드 및 request path 확인
- request param으로 전달 받은 값을 기준으로 Task 객체 가져와서 JSON으로 직렬화

MJbae committed:
feat: content와 http status code를 효과적으로 관리하기 위해 ResponseHeader 객체 생성

MJbae committed:
refactor: DemoHandler 내 content와 statusCode를 ResponseHeader 객체로 관리

MJbae committed:
refactor: method 이름 변경, extractTaskIdFromPath to extractTaskIdFrom

MJbae committed:
refactor: 문자열 구조를 쉽게 파악하기 위해 toString에 String::format 적용

MJbae committed:
fix: ResponseHeader 내 사용하지 않는 메소드 제거

MJbae commented:
작업 요약
- Task를 LinkedList로 관리하면 요소 검색 시 시간복잡도가 O(n)이라서 HashMap으로 관리하도록 수정했습니다.
- Task에 대한 조회 및 갱신하는 역할을 TaskRepository 객체로 옮겼습니다.
- 리뷰를 반영하여 메소드의 이름을 보다 간략하면서도 객체지향적으로 수정했습니다.

MJbae committed:
refactor: 객체의 역할을 모두 포괄하는 이름으로 변경(ResponseHeader -> HttpResponse)

MJbae committed:
feat: DemoHttpHandler 내 Task의 CRUD 책임을 담당할TaskRepository 클래스 생성

MJbae committed:
feat: Task 모델을 다루는 로직을 TaskRepository 클래스로 대체

MJbae committed:
refactor: 코드의 간결성을 위해 HttpResponse 객체의 역할을 DemoHandler::sendResponse가 대체

MJbae committed:
clean: 요구사항에 없는 PATCH 요청에 대한 처리로직 제거

MJbae committed:
clean: 불필요한 의존성 및 코드 제거

MJbae committed:
refactor: 객체지향적이지 않은 메소드 이름 수정(get 제거)

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,150 @@
+    private static final int HTTP_CREATE_CODE = 201;
+    private static final int HTTP_NO_CONTENT_CODE = 204;
+    private static final int HTTP_NOT_FOUND_CODE = 404;
+    private final TaskRepository repository;
```
위에 빈 줄 하나를 추가해서 상수들과 변수를 구분해 주면 좋을 것 같아요

MJbae replied:
구분해주는게 가독성 면에서 낫겠네요~ 감사합니다

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,150 @@
+        String path = httpRequestPath(exchange);
+        String body = httpRequestBody(exchange);
+
+        if (method == GET && path.equals("/tasks")) {
+            sendResponse(exchange, HTTP_OK_CODE, tasksToJson(mapper));
+        }
```
```suggestion
        if (method == GET && path.equals("/tasks")) {
            sendResponse(exchange, HTTP_OK_CODE, tasksToJson(mapper));
            return
        }
```

할 일이 끝나면 빠르게 종료하여, 다음 코드를 읽을 필요도 없이 만들어서, 코드를 읽는 사람에게 부하를 줄이면 좋을 것 같아요.

See also
  - https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html

MJbae replied:
동의합니다. 보호구문 개념을 적용하면 읽는 입장에서 더 편하겠네요~

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,150 @@
+            Task foundTask = repository.taskBy(taskId);
+
+            if (foundTask == null) {
+                sendResponse(exchange, HTTP_NOT_FOUND_CODE, "TaskId가 유효하지 않습니다");
```
유효하지 않다는 말은 해석될 수 있는 여지가 많습니다. 보안상의 이유로 일부러 정보를 숨기는 것이 아니라면, 구체적으로 이야기를 해주는게 좋을 것 같아요. 예를들어 유효하지 않다는 말은 taskId가 `abcd`로 요청온 것도 유효하지 않죠

MJbae replied:
생각해보니 매우 모호한 표현이네요. 구체적인 표현으로 바꾸는게 좋겠네요~

hannut91 commented on app/src/main/java/com/codesoom/assignment/TaskRepository.java:
```diff
@@ -0,0 +1,47 @@
+        return tasks.get(id);
+    }
+
+    public List<Task> tasksAll() {
+        return new ArrayList<>(tasks.values());
+    }
```
할 일 목록을 hashMap으로 관리하여, 찾거나 저장할 때 빠를 것 같아요. 만약 사용자가 등록한 순서대로 할 일 목록을 원한다면 어떻게 될까요?

MJbae replied:
오! 이펙티브 자바에서 언급한 Comparable Interface를 상속할 때가 왔군요! 
Task에서 Comparable을 상속해서 compareTo를 구현한 후, Task.id가 오름차순으로 자동 생성되니까 id를 기준으로 정렬되도록 작업하면 될 것 같네요

MJbae committed:
style: 가독성을 향상을 위해 상수와 변수 사이에 공백 추가

MJbae committed:
refactor: 코드 가독성을 높이기 위해 보호구문(guard clause) 활용

MJbae committed:
fix: 예외 원인을 구체적으로 전달하기 위해 예외 메시지 수정

MJbae committed:
refactor: 요구사항에 따라 생성한 순서를 기준으로 Task를 조회하도록 수정

작업내용
- Task 객체에서 Comparable Interface 상속 받아서 compareTo 메소드 구현
- Task 객체의 아이디는 오름차순으로 생성되므로 compareTo 메소드 내부적으로 id를 기준으로 정렬하도록 구현
- Task::compareTo를 적용하여 TaskRepository::tasksAll에서 생성순서에 따라 모든 Task를 조회할 수 있도록 수정

MJbae committed:
refactor: Task 객체를 간결하게 유지하기 위해 Comparable 인터페이스 제거하고 Comparator 인터페이스를 Respository에 적용

작업내용
- Comparable 인터페이스 및 compareTo 메소드를 Task 객체에서 제거
- TaskRepository::tasksAll에 사용되는 정렬메소드에 Comparator 인터페이스를 람다식으로 적용

MJbae committed:
refactor: handler 로직을 간소화하기 위해 request param으로 전달한 taskId가 없을 경우 예외처리

MJbae committed:
fix: 예외 메시지를 전달하기 위해 Exception::getMessage로 대체

MJbae committed:
refactor: 코드의 간결함과 유지보수 용이를 위해 JSON 직렬화 역직렬화의 책임 TaskMapper로 이동

작업내용
- Object <-> JSON 간 직렬화 역직렬화 책임을 TaskMapper로 이동
- App 최상위계층에서 Mapper를 주입 받도록 설계
- DemoHandler 내 직렬화 역직렬화 메소드 제거

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,128 @@
+            }
+
+            if (method == POST && path.equals("/tasks")) {
+                Task newTask = mapper.toTask(body);
+                newTask = repository.save(newTask);
```
`newTask`를 변수로 사용하고 계신 것 같아요. 변수가 변경될 수 있는 변수로 사용하기 보다는, 상수로 사용하는게 좋습니다. 중간에 변수가 변경될 수 있다는 사실 하나로 복잡성이 많이 증가합니다

MJbae replied:
그렇군요! 멀티쓰레딩 환경에서 중간에 값이 바뀔 수 있겠네요. `final Task newTask`로 선언하여 중간에 값이 바뀌는 것을 막아보겠습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,128 @@
+
+            if (method == DELETE && path.startsWith("/tasks")) {
+                long taskId = extractTaskIdFrom(path);
+                repository.taskBy(taskId);
```
이 것은 필요 없어보이네요

MJbae replied:
TaskRepository::update, TaskRepository::delete 내부에서 재사용될 수 있는 메소드를 Handler에서 사용하고 있네요. 옮기는게 좋겠네요~

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,128 @@
+                sendResponse(exchange, HTTP_NO_CONTENT_CODE, "정상적으로 삭제되었습니다");
+            }
+        } catch (NoSuchElementException e) {
+            sendResponse(exchange, HTTP_NOT_FOUND_CODE, "taskId에 해당하는 리를 찾을 수 없습니다 from handler layer \n"
+                    + e.getMessage() + " from repository layer");
```
이 에러메시지를 누가 보는지 생각해 보시면, 에러메시지를 어떻게 작성하는게 좋을지 많은 힌트가 될거에요

MJbae replied:
sendResponse의 수신자가 보기에 repository layer에 대한 예외 메시지는 과한 면이 있어 보이네요.

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,128 @@
+import static com.codesoom.assignment.HttpMethod.*;
+
+public class DemoHttpHandler implements HttpHandler {
+    private static final int HTTP_OK_CODE = 200;
+    private static final int HTTP_CREATE_CODE = 201;
+    private static final int HTTP_NO_CONTENT_CODE = 204;
+    private static final int HTTP_NOT_FOUND_CODE = 404;
```
## 

HTTP 프로토콜의 상태값을 상수로 잘 선언하셨습니다. 그런데 앞부분의 `HTTP_`가 반복적으로 나타나는 것이 아쉬우셨을 것 같아요. 그러니 이 중복을 제거해 보도록 합시다. 이 값들을 `enum`으로 다른 파일에 정의해 보세요. 이번 기회에 RFC 7231을 함께 읽어보셔도 좋은 공부가 될 거에요.

[https://datatracker.ietf.org/doc/html/rfc7231#section-6](https://datatracker.ietf.org/doc/html/rfc7231#section-6)

MJbae replied:
HttpStatus라는 객체를 정의하고 HttpStatus에게 Code를 정중하게 요청하는 방식으로 수정해보았습니다. 
코드도 간결해지고 HTTP Status Code에 대한 관리포인트도 명확해지니 참 좋네요~

MJbae committed:
refactor: 외부에 의해 변수의 값이 변경되는 것을 방지하기 위해 변수의 상수화

작업내용
- Task newTask로 선언되고 중간에 값이 바뀌었지만 final 선언을 함으로써 상수화하여 변수 값이 바뀌는 것을 방지

MJbae committed:
refactor: hanlder에서 불필요하게 사용되고 있는 TaskRepository::taskBy 메소드 제거

작업내용
- Handler에서 TaskRepository::taskBy 메소드 제거
- Repository::update, Repository::delete에서 taskBy 재사용하는 방법으로 대체

MJbae committed:
refactor: HTTP Status Code에 대해 상수에서 Enum으로 대체

MJbae committed:
fix: sendResponse로 보내는 메시지를 수신자에 맞게 수정

MJbae committed:
refactor: 중복 제거에 따른 관리포인트를 줄이기 위해 taskToJson과 tasksToJson 수정

작업내용
- toJson의 매개변수를 Object로 수정하여 모든 객체에 대해 처리하도록 수정
- tasksToJson에서 toJson 재사용하도록 수정

MJbae committed:
refactor: hanlde 내부 로직을 메소드로 분리

MJbae committed:
refactor: 코드의 간결함을 위해 path가 tasks 끝나는 부분과 task로 끝나는 부분을 메소드로 분리

MJbae committed:
refactor: 코드 중복을 줄이기 위해 task를 추출하는 공통 로직을 밖으로 추출

MJbae committed:
refactor: path param으로 전달한 taskId가 없는 경우에 예외처리를 handleItem으로 이동

MJbae commented:
윤석님의 리뷰와 아샬님의 풀이 영상을 참고해서 리팩토링 작업을 했습니다!

계속 수정하면서 코드가 많이 간결해진 것 같아서 뿌듯하네요 ㅎㅎ
